great_expectations/core/config_substitutor.py

```diff
@@ -0,0 +1,352 @@
+import base64
+import json
+import logging
+import re
+from collections import OrderedDict
+from functools import lru_cache
+from typing import Optional
+
+import great_expectations.exceptions as ge_exceptions
+from great_expectations.data_context.types.base import BaseYamlConfig
+
+try:
+    from azure.identity import DefaultAzureCredential
+    from azure.keyvault.secrets import SecretClient
+except ImportError:
+    SecretClient = None
+    DefaultAzureCredential = None
+
+try:
+    import boto3
+    from botocore.exceptions import ClientError
+except ImportError:
+    boto3 = None
+    ClientError = None
+
+try:
+    from google.cloud import secretmanager
+except ImportError:
+    secretmanager = None
+
+logger = logging.getLogger(__name__)
+
+
+class _ConfigurationSubstitutor:
+    """
+    TODO
+    """
+
+    def __init__(self) -> None:
+        # Using the @lru_cache decorator on method calls can create memory leaks - an attr is preferred here.
+        # Ref: https://stackoverflow.com/a/68550238
```

Comment:
This was added by me - everything else in this file while directly ported over from `util.py`

Reply:
Happy to clean up the file (refactors and type hints) but perhaps that should be a subsequent effort